File: WFM-Planning-Suite/web/src/core/simulation/des.ts
import type { SimulationResult, SizingResult } from '../models'

function seededRandom(seed: number) {
  let s = seed
  return () => { s = (s * 1664525 + 1013904223) & 0xffffffff; return (s >>> 0) / 0xffffffff }
}

export function runSimulation(
  sizing: SizingResult, aht: number, patience: number, agents: number, warmup = 5, seed = 42,
): SimulationResult {
  const rand = seededRandom(seed)
  const intervalSec = sizing.profile.intervalMinutes * 60
  let answered = 0, abandoned = 0, withinSla = 0, totalWait = 0, intervalsSimulated = 0

  for (let idx = 0; idx < sizing.rows.length; idx++) {
    if (idx < warmup) continue
    intervalsSimulated++
    const volume = Math.round(sizing.rows[idx].volume)
    if (volume <= 0) continue

    const calls = Array.from({ length: volume }, () => ({
      arrival: rand() * intervalSec,
      patience: patience > 0 ? -Math.log(1 - rand()) * patience : Infinity,
      service: aht > 0 ? -Math.log(1 - rand()) * aht : 1,
    })).sort((a, b) => a.arrival - b.arrival)

    const agentFree = Array(Math.max(1, agents)).fill(0)
    for (const call of calls) {
      let best = 0
      for (let i = 1; i < agentFree.length; i++) if (agentFree[i] < agentFree[best]) best = i
      let wait = 0
      if (agentFree[best] > call.arrival) {
        wait = agentFree[best] - call.arrival
        if (wait > call.patience) { abandoned++; continue }
      }
      const start = Math.max(call.arrival, agentFree[best])
      totalWait += wait
      agentFree[best] = start + call.service
      answered++
      if (wait <= sizing.profile.slaTimeSeconds) withinSla++
    }
  }

  const total = answered + abandoned
  return {
    slaPct: total > 0 ? (withinSla / total) * 100 : 100,
    asaSeconds: answered > 0 ? totalWait / answered : 0,
    abandonmentPct: total > 0 ? (abandoned / total) * 100 : 0,
    intervalsSimulated,
    crossCheckPassed: null,
    crossCheckNotes: null,
  }
}

export function crossValidate(sizing: SizingResult, sim: SimulationResult, tolerance = 15): SimulationResult {
  if (!sizing.rows.length) {
    return { ...sim, crossCheckPassed: true, crossCheckNotes: 'No sizing rows to compare' }
  }
  const avgSla = sizing.rows.reduce((a, r) => a + r.slaPct, 0) / sizing.rows.length
  const delta = Math.abs(avgSla - sim.slaPct)
  return {
    ...sim,
    crossCheckPassed: delta <= tolerance,
    crossCheckNotes: `Analytic avg SLA ${avgSla.toFixed(1)}% vs simulation ${sim.slaPct.toFixed(1)}% (delta ${delta.toFixed(1)}%)`,
  }
}

File: WFM-Planning-Suite/src/core/simulation/des.py
"""Discrete-event simulation for cross-validation against analytic Erlang sizing."""

from __future__ import annotations

import random
from dataclasses import dataclass

from core.models import SimulationResult, SizingResult


@dataclass
class _Call:
    arrival: float
    patience: float
    service_time: float


def run_simulation(
    sizing: SizingResult,
    aht_seconds: float,
    patience_seconds: float,
    agents: int,
    warmup_intervals: int = 5,
    rng_seed: int = 42,
) -> SimulationResult:
    rng = random.Random(rng_seed)
    interval_seconds = sizing.profile.interval_minutes * 60

    total_answered = 0
    total_abandoned = 0
    total_within_sla = 0
    total_wait = 0.0
    intervals_simulated = 0

    for idx, row in enumerate(sizing.rows):
        if idx < warmup_intervals:
            continue
        intervals_simulated += 1
        volume = int(round(row.volume))
        if volume <= 0:
            continue

        calls: list[_Call] = []
        for _ in range(volume):
            arrival = rng.uniform(0, interval_seconds)
            patience = rng.expovariate(1.0 / patience_seconds) if patience_seconds > 0 else float("inf")
            service = rng.expovariate(1.0 / aht_seconds) if aht_seconds > 0 else 1.0
            calls.append(_Call(arrival=arrival, patience=patience, service_time=service))
        calls.sort(key=lambda c: c.arrival)

        agent_free = [0.0] * max(1, agents)

        for call in calls:
            best = min(range(len(agent_free)), key=lambda i: agent_free[i])
            if agent_free[best] <= call.arrival:
                start = call.arrival
                wait = 0.0
            else:
                wait = agent_free[best] - call.arrival
                if wait > call.patience:
                    total_abandoned += 1
                    continue
                start = agent_free[best]

            total_wait += wait
            agent_free[best] = start + call.service_time
            total_answered += 1
            sla_time = sizing.profile.sla_time_seconds
            if wait <= sla_time:
                total_within_sla += 1

    total = total_answered + total_abandoned
    sla_pct = (total_within_sla / total * 100.0) if total > 0 else 100.0
    asa = (total_wait / total_answered) if total_answered > 0 else 0.0
    abandon_pct = (total_abandoned / total * 100.0) if total > 0 else 0.0

    return SimulationResult(
        sla_pct=sla_pct,
        asa_seconds=asa,
        abandonment_pct=abandon_pct,
        intervals_simulated=intervals_simulated,
        warmup_excluded=True,
    )


def cross_validate(sizing: SizingResult, simulation: SimulationResult, tolerance_pct: float = 15.0) -> SimulationResult:
    if not sizing.rows:
        simulation.cross_check_passed = True
        simulation.cross_check_notes = "No sizing rows to compare"
        return simulation

    avg_analytic_sla = sum(r.sla_pct for r in sizing.rows) / len(sizing.rows)
    delta = abs(avg_analytic_sla - simulation.sla_pct)
    passed = delta <= tolerance_pct
    simulation.cross_check_passed = passed
    simulation.cross_check_notes = (
        f"Analytic avg SLA {avg_analytic_sla:.1f}% vs simulation {simulation.sla_pct:.1f}% "
        f"(delta {delta:.1f}%, tolerance {tolerance_pct}%)"
    )
    return simulation
